traders/fsm.py

```diff
@@ -0,0 +1,170 @@
+        """Returns the next state without updating it"""
+        return self._next_state
+
+    def apply(self):
+        """Applies the update function and returns the new state"""
+        if self._update is not None:
+            self._update(self.current_state, self.event, self._next_state)
+        return self._next_state
+
+    def match(self, current_state, event):
+        """Verifies if the transition matches the state and event"""
+
+        if self.current_state == current_state and self.event == event:
+            return True
+
+        return False
+
+
+class FiniteStateMachine(State):
+
+    def __init__(self, name, start_state):
+        """A finite state machine
+
+        The finite state machine encapsulates a list of states and the
+        transition between them. Because a FSM is itself a state, it can be
+        run and yield events. The events of a FSM are all the events of its
+        state which are not associated with a transition.
+
+        """
+
+        super().__init__(name)
+        self.start_state = start_state
+        self._transitions = []
+
+    @property
+    def events(self):
+        """Returns the list of events that can be thrown by the  FSM"""
+
+        # The events that the FSM can throw are all the events that
+        # do not have a transition.
+        events = []
+        states = set(chain((self.start_state,),
+                           (t.current_state for t in self._transitions),
+                           (t.next_state for t in self._transitions)))
+        for state in states:
+            for event in state.events:
+                if self.transition(state, event) is None:
+                    events.append(event)
+
+        return tuple(events)
+
+    def add(self, transition):
+        """Adds a transition to a FiniteStateMachine"""
+
+        if not isinstance(transition, Transition):
+            raise TypeError(
+                '\'transition\' must be an instance of {}, not {}.'
+                .format(Transition, transition.__class__))
+        self._transitions.append(transition)
+
+    def pretty_print(self):
+        """Pretty prints the FSM for debugging"""
+
+        out = 'FSM: {}\n'.format(self.name)
+        out += '  Start state: {}\n'.format(str(self.start_state))
+        out += '  Trantisions:\n'
+        for transition in self._transitions:
+            out += '    {}\n'.format(str(transition))
+
+        print(out)
+
+    def run(self):
+        """Runs the finite state machine"""
+
+        current_state = self.start_state
+
+        while True:
+            event = current_state.run()
+            transition = self.transition(current_state, event)
+            if transition is None:
+                break
+            current_state = transition.apply()
+
+        return event
+
+    def transition(self, state, event):
+        """Returns the transition for a state and event or None"""
+        matches = (t for t in self._transitions if t.match(state, event))
```

Comment:
Since now we have the name of the states, I think we should create a dictionary(str, str) -> Transition using the name of the state and the event.